web/src/helpers/api.js

```diff
@@ -0,0 +1,86 @@
+import * as request from 'd3-request';
+import { sha256 } from 'js-sha256';
+import Cookies from 'js-cookie';
+
+import { isLocalhost, isProduction } from './environment';
+import thirdPartyServices from '../services/thirdparty';
+
+function isRemoteParam() {
+  return (new URLSearchParams(window.location.search)).get('remote') === 'true';
+}
+
+// Use local endpoint only if ALL of the following conditions are true:
+// 1. The app is running on localhost
+// 2. The `remote` search param hasn't been explicitly set to true
+// 3. Document domain has a non-empty value
+function isUsingLocalEndpoint() {
+  return isLocalhost() && !isRemoteParam() && document.domain !== '';
+}
+
+export function getEndpoint() {
+  return isUsingLocalEndpoint() ? 'http://localhost:9000' : 'https://api.electricitymap.org';
+}
+
+export function protectedJsonRequest(path) {
+  const url = getEndpoint() + path;
+  const token = isUsingLocalEndpoint() ? 'development' : ELECTRICITYMAP_PUBLIC_TOKEN;
+  const timestamp = new Date().getTime();
+
+  return new Promise((resolve, reject) => {
+    request.json(url)
+      .header('electricitymap-token', Cookies.get('electricitymap-token'))
+      .header('x-request-timestamp', timestamp)
+      .header('x-signature', sha256(token + path + timestamp))
+      .get(null, (err, res) => {
+        if (err) {
+          reject(err);
+        } else if (!res || !res.data) {
+          reject(new Error(`Empty response received for ${url}`));
+        } else {
+          resolve(res.data);
+        }
+      });
+  });
+}
+
+export function textRequest(path) {
+  const url = getEndpoint() + path;
+
+  return new Promise((resolve, reject) => {
+    request.text(url).get(null, (err, res) => {
+      if (err) {
+        reject(err);
+      } else if (!res) {
+        reject(new Error(`Empty response received for ${url}`));
+      } else {
+        resolve(res);
+      }
+    });
+  });
+}
+
+function trackError(err, appState) {
```

Comment:
this should probably be in a different helper?

Reply:
Good catch, I moved it into `thirdPartyServices` in https://github.com/tmrowco/electricitymap-contrib/pull/2388/commits/dbd8406a638090ef773ee5f480d939da8e7b9b1d which also simplified the sagas a bit since the app state didn't need to be carried across :+1: